src/MergeContext.js

```diff
@@ -108,6 +108,96 @@ class Approval {
     }
 }
 
+class StatusCheck
+{
+    constructor(context, state, targetUrl, description) {
+        assert(context);
+        assert(state);
+        assert(targetUrl);
+        assert(description);
+
+        this.context = context;
+        this.state = state;
+        this.targetUrl = targetUrl;
+        this.description = description;
+    }
+}
+
+// Passed status checks analysis
+class StatusResult
+{
+    // checksNumber:
+    //   for staged commits: the bot-configured number of required checks (Config.stagingChecks()),
+    //   for PR commits: GitHub configured number of required checks (requested from GitHub)
+    // context: either "PR" or "Staging";
+    constructor(checksNumber, context) {
+        assert(checksNumber !== undefined && checksNumber !== null);
+        assert(context);
+        this.checksNumber = checksNumber;
+        this.context = context;
+        this.statusChecks = [];
+    }
+
+    addStatus(statusCheck) {
+        assert(statusCheck);
+        this.statusChecks.push(statusCheck);
+    }
+
+    // there are some pending contexts except contextName
+    othersPending(contextName) {
+        if (!this.pending())
+            return false;
+        if (this.statusChecks.find(check => check.context === contextName && check.state === 'pending') !== undefined)
+            return false;
+        return true;
+    }
+
+    // there is only one pending context contextName
+    singlePending(contextName) {
+        return this.pending() && !this.othersPending(contextName);
+    }
+
+    // whether at we have at least checksNumber checks and some of them are pending
+    pending() {
+        if (this.statusChecks.length < this.checksNumber)
+            return true;
+        if (this.statusChecks.find(check => check.state === 'pending'))
+            return true;
+        return false;
+    }
+
+    // whether some of checks failed
+    failed() {
+        return this.statusChecks.find(check => check.state === 'failure' || check.state === 'error') !== undefined;
+    }
+
+    // whether at least checksNumber checks finished and none of them failed
+    succeeded() {
+        if (this.pending() || this.failed())
```

Comment:
After the changes discussed elsewhere in this review, and after making the implementation more reliable/robust, this method would look similar to the following:

```
// the results are final and all checks were a success
succeeded() {
    return this.final() && !this.statusChecks.find(check => check.state !== 'success');
}
```